libs/langchain/langchain/retrievers/web_research.py

```diff
@@ -0,0 +1,169 @@
+from langchain.utilities import GoogleSearchAPIWrapper
+from langchain.vectorstores.base import VectorStore
+
+logger = logging.getLogger(__name__)
+
+
+class SearchQueries(BaseModel):
+    """Search queries to run to research for the user's goal."""
+
+    queries: List[str] = Field(
+        ..., description="List of search queries to look up on Google"
+    )
+
+
+DEFAULT_SEARCH_PROMPT = PromptTemplate(
+    input_variables=["question"],
+    template="""<<SYS>> \n You are a web research assistant to help users
+    answer questions. Answer using a numeric list. Do not include any extra
+    test. \n <</SYS>> \n\n [INST] Given a user input search query, 
+    generate a numbered list of five search queries to run to help answer their 
+    question: \n\n {question} [/INST]""",
+)
+
+
+class LineList(BaseModel):
+    """List of questions."""
+
+    lines: List[str] = Field(description="Questions")
+
+
+class QuestionListOutputParser(PydanticOutputParser):
+    """Output parser for a list of numbered questions."""
+
+    def __init__(self) -> None:
+        super().__init__(pydantic_object=LineList)
+
+    def parse(self, text: str) -> LineList:
+        lines = re.findall(r"\d+\..*?\n", text)
+        return LineList(lines=lines)
+
+
+class WebResearchRetriever(BaseRetriever):
+    # Inputs
+    vectorstore: VectorStore = Field(
+        ..., description="Vector store for handling document embeddings"
+    )
+    llm: Union[BaseLLM, ChatOpenAI] = Field(
+        ..., description="Language model for generating questions"
+    )
+    GOOGLE_CSE_ID: str = Field(..., description="Google Custom Search Engine ID")
+    GOOGLE_API_KEY: str = Field(..., description="Google API Key")
+    search_prompt: PromptTemplate = Field(
+        DEFAULT_SEARCH_PROMPT, description="Search Prompt Template"
+    )
+    max_splits_per_doc: int = Field(100, description="Maximum splits per document")
+    num_search_results: int = Field(1, description="Number of pages per Google search")
+
+    def search_tool(self, query: str, num_search_results: int = 1) -> List[dict]:
+        """Returns num_serch_results pages per Google search."""
+        try:
+            os.environ["GOOGLE_CSE_ID"] = self.GOOGLE_CSE_ID
+            os.environ["GOOGLE_API_KEY"] = self.GOOGLE_API_KEY
+            search = GoogleSearchAPIWrapper()
+        except Exception as e:
+            print(f"Error: {str(e)}")
+        result = search.results(query, num_search_results)
+        return result
+
+    def _get_relevant_documents(
+        self,
+        query: str,
+        *,
+        run_manager: CallbackManagerForRetrieverRun,
+    ) -> List[Document]:
+        """Search Google for documents related to the query input.
+
+        Args:
+            query: user query
+
+        Returns:
+            Relevant documents from all various urls.
+        """
+
+        # Get search questions
+        logger.info("Generating questions for Google Search ...")
+        llm_chain = LLMChain(
```

Comment:
the llm_chain should be an attribute on this class, can be constructed from a class method